2019-12-naming-part1/naming.md

```diff
@@ -0,0 +1,231 @@
+
+## Naming functions
+
+Communication is most effective when it spans various levels of abstraction, when it travels up and down the ladder. Just stating that you should consider abstract ideas when naming things in your code is not enough. Providing abstract naming advice also requires concrete examples in order to be convincing. Concrete examples explain abstract ideas. In order to explain the color green, you might use examples like leaves in summertime, the bottom light of a traffic signal, or the color of an emerald.
+
+Consider a requirement in your application to display a person's first and last name. Simple enough. But the name you retrieve, and all other information associated with a person, comes from an external data source that you do not own. Sometimes the data contains extra whitespace &mdash; leading, trailing, and in between. For example, the name might arrive in your application as "   Jane  Doe ". The unpredictable, extra spaces cause issues in your UI and you need to strip the whitespace out for presentation.
+
+Since you do not own the data source, you cannot fix the problem at its root. The example string above contains extra whitespace before, after, and between individual parts of the name. Simply trimming leading and trailing whitespace, as most `trim` functions do, will not be enough &mdash; it will leave extra whitespace in the middle of the string. You need to isolate the words in the name and join them with a space. `String.split` and `Enum.join`, used together, do exactly what you want.
+
+```elixir
+# In syntax familiar to most programmers
+name = "   Jane  Doe "
+list = String.split(name) # => ["Jane", "Doe"]
+Enum.join(list, " ") # => "Jane Doe"
+```
+
+Since these examples are being shown in Elixir, henceforth, they will use idiomatic Elixir syntax.[^1] Elixir's pipe operator `|>` passes the result of one operation as the first argument to the next. It affords graceful and easy-to-read function composition:
+
+```elixir
+# Returns "Jane Doe"
+"   Jane  Doe "
+|> String.split()
+|> Enum.join(" ")
+```
+
+### Bottom rung: dirty mirrors
+
+You have an implementation that works. You have, of course, written tests (not shown). In your application, there's a `PersonPresenter` module, in which this functionality belongs. But, now you must create a function and name it.
+
+```elixir
+defmodule PersonPresenter do
+  # literal function name
+  def split_and_join(name)
+    name
+    |> String.split()
+    |> Enum.join(" ")
+  end
+end
+```
+
+At this point, it might be tempting to reach for an overtly literal function name, like `split_and_join` above. Indeed, the code splits and joins a name. It's quick and easy to conceive of; it's convenient.
```

Comment:
This was the only spot that distracted me. I think the semicolon is too jarring for this sentence It's quick and easy to conceive of; it's convenient. I might prefer "It's quick and easy to conceive of -- it's a convenient option." Might be good to emphasize that it's a little too easy to devise. It's the lazy choice.

Reply:
Addressed in bb7a6a3a9ed344ecafaedf8ebe2ea5d183cf6f35